crates/core/src/module/watch_graph.rs

```diff
@@ -0,0 +1,220 @@
+use std::hash::Hash;
+
+use hashbrown::{HashMap, HashSet};
+use petgraph::{
+  graph::{DefaultIx, NodeIndex},
+  stable_graph::StableDiGraph,
+  EdgeDirection,
+};
+
+use crate::error::{CompilationError, Result};
+
+/// ```md
+///     a
+///    /
+///  v_b
+/// ```
+#[derive(Debug, Hash, PartialEq, Eq)]
+enum EdgeMode {
+  /// a -> v_b
+  RootImport,
+  /// a <- v_b
+  WatchImport,
+}
+
+impl EdgeMode {
+  fn is_root_import(&self) -> bool {
+    *self == Self::RootImport
+  }
+}
+
+pub struct WatchGraph {
+  g: StableDiGraph<String, EdgeMode>,
+  id_index_map: HashMap<String, NodeIndex<DefaultIx>>,
+}
+
+impl WatchGraph {
+  pub fn new() -> Self {
+    Self {
+      g: StableDiGraph::new(),
+      id_index_map: HashMap::new(),
+    }
+  }
+
+  pub fn add_node(&mut self, node: String) {
+    if !self.has_module(&node) {
+      let index = self.g.add_node(node.clone());
+      self.id_index_map.insert(node, index);
+    }
+  }
+
+  pub fn add_edge(&mut self, from: &String, to: &String) -> Result<()> {
+    let from_index = self.id_index_map.get(from).ok_or_else(|| {
+      CompilationError::GenericError(format!(
+        r#"from node "{}" does not exist in the module graph when add edge"#,
+        from
+      ))
+    })?;
+
+    let to_index = self.id_index_map.get(to).ok_or_else(|| {
+      CompilationError::GenericError(format!(
+        r#"to node "{}" does not exist in the module graph when add edge"#,
+        to
+      ))
+    })?;
+    //         a                          h               c
+    //       /   \                      /               /
+    //     b      v_c                v_e              v_e
+    //   /       /   \
+    // v_d     v_e   v_f
+    //           \    /
+    //            v_g
+    // e change causes a、f、c update
+    // c change causes a
+    // g change causes a update
+    //
+    // m  |  root
+    // a -> []
+    // b -> []
+    // c -> [d]
+    // d -> [b]
+    // e -> [a,h,c]
+    // f -> [a]
+    // g -> [a]
+    // h -> []
+    let roots: Vec<String> = self
+      .relation_roots(from)
+      .into_iter()
+      .map(|item| item.clone())
+      .collect();
+
+    let roots: Vec<String> = [roots, vec![from.clone()]].concat();
+
+    for root in roots.into_iter() {
+      let root_id = self.id_index_map.get(&root).unwrap();
+
+      self
+        .g
+        .update_edge(*to_index, *root_id, EdgeMode::RootImport);
+    }
+
+    self
+      .g
+      .update_edge(*from_index, *to_index, EdgeMode::WatchImport);
+
+    Ok(())
+  }
+
+  pub fn resources(&self) -> Vec<&String> {
```

Comment:
Rename `resources` to `modules`, `resources` is refer to the production files in Farm